### DISCUSSION TITLE
Definition and Non-Linearity of Models

### DISCUSSION THREAD

**Discussion by michaelweinold** [Depart-de-Sentier/sentier.dev#19] [General]
From what I understand, the main point of this project is a common glossary and associated infrastructure. However, I have some question about the product system models (PSMs), as they are currently envisioned in the Product Vision:

<img width="450" alt="Screenshot 2024-04-15 at 20 28 46" src="https://github.com/Depart-de-Sentier/sentier.dev/assets/23102087/7c22b073-cab7-4753-9ed9-63ff53a7d7ae">

These are supposed to be _general, possibly non-linear_ production functions, in practice derived from Excel/VBA/Python/... models of "heroes" (domain experts), generally described by:

$f(i_1, i_2, \dots) = o, b_1, b_2, \dots$

where

$i \dots$ (technosphere/biosphere) input
$o \dots$ (technosphere) output
$b \dots$ environmental burden (=biosphere output)

The Product Vision also mentions, that:

> Models are the logical rules explaining relationships between inputs and outputs (...). By separating these models from the data store, the model can search for the data it needs on demand. This search can be context-aware, looking only for the time, place, or specific product needed. (...)

Following discussions on Saturday, 13th of April, I have one primary question:

Are the models valid only in a specific geographic, temporal context? If so, why would they look for data outside this context? Or would all models be so generic, that they describe the physics/chemistry of production processes _everywhere_ and all that matters is "where the input comes from"?

I would imagine that many "models" are build from observations (eg. data provided by industry). In that case, they would be specific to input data of their spatial/temporal context.

...and some practical questions:

 - [ ] Is non-linearity in the product system models tangible benefit?
 I know @romainsacchi [has recently published on this](https://doi.org/10.3389/frsus.2020.611593) - although that was for the foreground only (?).
 - [ ] If we can use linear models, could the PSMs build columns in the A-matrix "on-demand", whenever the correct context has been determined?
 - [ ] Do we know how we would mathematically deal with loops in the graph? See also   https://github.com/Depart-de-Sentier/sentier.dev/discussions/16.

**Comment by cmutel** [comment]
> Are the models valid only in a specific geographic, temporal context?

The models should be simplifications of the fundamental physics/biology/chemistry/etc. of the industrial economy. As a general answer, this should not vary over time and space.

In practice, we do see the same technology being operated differently over time and space, and different technologies in different times and places. But, at least IMHO, this is not a problem of a well-formed model - rather, to accurately reflect this variability we need the best input data we can get. This *input data* is also gathered in a spatial and temporal context, and can be quantitative or categorical (e.g. presence of specific pollution abatement technology). As such, there is no fundamental problem with creating generic models.

That being said, my guess is that we will have some models which are restricted to certain places or time frames. In the process of going from complex realities to simplified models, we make choices based on our social and political context, the number and types of measurements available, our training and experience, etc. And it is often the case that the available data and models are tightly correlated. We want to publish the best available information, which probably means region-specific models, which in turn means some non-generic models.

BTW, the "heroes" are the people doing maintenance and community support - often but not necessarily the model developers themselves.

**Comment by cmutel** [comment]
> Do we know how we would mathematically deal with loops in the graph?

We can come back to this now, but my assumption going forward is that such loops will tail off into small numbers quite quickly. If there are cases where those loops don't have this quickly decreasing behaviour (such as eco-parks which reuse each other's wastes), then these are best handled together in a single model where there are more appropriate mathematical approaches.

**Comment by cmutel** [comment]
>  If we can use linear models, could the PSMs build columns in the A-matrix "on-demand"

I think we might be overloading the linear versus non-linear discussion. I don't really care if a model's response surface is a plane or something more interesting - we aren't going in this direction because we like playing with fun algorithms. I mean, we do, but that isn't the main motivation. Instead, the point of building models instead of using linear coefficients is to better reflect how systems perform by forcing the modeller to identify the independent and dependent parameters. Reducing the set of parameters to these two classes leads to better uncertainty analysis results (and hence better decision support), easier data collection (because the number of parameters is reduced), *much better and more consistent* maintainability (because the next person doesn't have to figure out how the different values were related), and easier adaption to specific contexts or ranges of input parameters.

Of course we can calculate the full set of parameters and create a process in the **A** matrix. The degree of linearity in the response surface will determine the number of processes we need to generate from the model to accurately represent its possible outputs.

**Comment by michaelweinold** [comment]
@cmutel, since this was recently promoted in a Brightway mailing list:

How would [`lca_algebraic`](https://github.com/oie-mines-paristech/lca_algebraic) fit into this? Not being too familiar with parametric LCA, it sounds like... a way to write out PSMs using symbolic Python?

**Reply by cmutel** [reply]
`lca_algebraic` is a wrapper on top of [Sympy](https://www.sympy.org/en/index.html) - as such, it has a very strong way of describing systems. Sympy is a computer algebra system like Mathematica, and main idea of Sympy is to define variables in Python, and then to write equations like you would on paper.

We definitely need something like this, and I will look into it when we start working more intensely on models. One thing I don't know enough about is how this interacts with DataFrames which have correlated samples or observations.
